### ISSUE TITLE
Circular type false positive with async/await 

### ISSUE THREAD

**Issue by jeremy-rifkin** [microsoft/TypeScript#51115]
# Bug Report

### 🔎 Search Terms

implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer

await

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about "referenced directly or indirectly"

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.4#code/IYZwngdgxgBAZgV2gFwJYHsIwOYFNkD6yAFqhNiABQEBcMECAtgEa4BOMAPjEgCa5wyuXgEoYAbwCwAKBhyYbfAjZYA2gAYAugG4ZAXxkzQkWIhQYscdOmbA2lMVNnyANvhisriugxbsuPBD8ghDCMAC8gcFCvLrOcl6UbsgwqBEw6tqpMAA8MABMWagA1MWOMvKVMFCYICmKIAguyCDpwADuwKgpeIQkZBSUnuiKInFV8sOK6Q1NLRo6FfIG0itAA)

### 💻 Code

```ts
async function get_things(_: number | undefined) {
    return [0];
}

async function foobar() {
    // The error goes away without the `| undefined`
    let before: number | undefined = undefined;
    // The error goes away without the loop
    for(let i = 0; i < 2; i++) {
        // The error goes away without the async/await
        const results = await get_things(before);
        //       ^ 'results' implicitly has type 'any' because it does not have a type annotation
        //          and is referenced directly or indirectly in its own initializer.
        before = results[0];
    }
}

```

### 🙁 Actual behavior

This results in an error
```
'results' implicitly has type 'any' because it does not have a type annotation
and is referenced directly or indirectly in its own initializer.
```

### 🙂 Expected behavior

This should not be an error.

**Comment by jeremy-rifkin** [comment]
This bug is similar to https://github.com/microsoft/TypeScript/issues/48708 but is present as far back as the playground goes (3.3.3). It stands apart largely because of the async/await.

**Comment by whzx5byb** [comment]
Duplicate of https://github.com/microsoft/TypeScript/issues/43047

**Comment by Andarist** [comment]
FYI: I have a fix locally for this one - just need to clean the code up and prepare a PR some time later
